### ISSUE TITLE
Named tensors

### ISSUE THREAD

**Issue by JuliusKunze** [google/jax#5048]
PyTorch has experimental support for [named tensors](https://pytorch.org/docs/stable/named_tensor.html) achieving [some compelling design goals](http://nlp.seas.harvard.edu/NamedTensor) while keeping existing code compatible. For example, binop broadcasting is still based on dimension order ([unlike in xarray](https://xarray.pydata.org/en/stable/computation.html#broadcasting-by-dimension-name)), consistent with standard NumPy/JAX/... semantics, but checks that aligned dimension names match.

It would be great to have named tensors that work both in op-by-op and under function transformations in JAX.

@shoyer In https://github.com/google/jax/issues/1565 you mentioned that this could be done by wrapping JAX based on https://github.com/google/jax/pull/611. According to my current understanding, this means:
- Add name rules for lax primitives, returning the output dimension names for given input dimension names.
- Add a corresponding `eval_names` transform.
- Add a `NamedDeviceArray` subtype of `DeviceArray` that adds a `names` property.
- We want names to be propagated in op-by-op mode on `NamedDeviceArray`s. For that,
    - make a named version of `jax.numpy`, wrapping each op with the `named` transform. 
    - attach it to `NamedDeviceArray` using https://github.com/google/jax/pull/611 (+1 for merging). Alternatively, one could rewrite `jax.numpy` using `numpy_dispatch.get_array_module` from https://github.com/google/jax/pull/4076 (appears cumbersome).
- I guess one would also need to make a named version of `jit` since feeding a `NamedDeviceArray` into a `jit`ted function would return a `DeviceArray` at best (similar `grad`, `vmap` etc.).

Is this plan sound? @shoyer @mattjj Would you update (and merge, if successful) https://github.com/google/jax/pull/611 just for this application? In that case, I'd be interested in prototyping a named tensor library for JAX, with a good amount of passion, in accordance with https://github.com/google/jax/issues/1565. (:

**Comment by Jeevesh8** [comment]
Have you started working on this @JuliusKunze ?

**Comment by apaszke** [comment]
We are actually working on something that will pretty much realize the plan that @JuliusKunze has outlined here, with some additional benefits too (e.g. making it very easy to shard those programs with named axes over multiple accelerators).

**Comment by JuliusKunze** [comment]
@Jeevesh8 No, and now I won't anymore. (: @apaszke That's great to hear! Will this go into the JAX repo?

**Comment by degregat** [comment]
Do I assume correctly that this evolved into [named axes](https://jax.readthedocs.io/en/latest/notebooks/xmap_tutorial.html), or is there another module I did not find?

**Comment by froystig** [comment]
That's correct.

**Comment by juliuskunze** [comment]
@apaszke @froystig That looks awesome! Rad choice not taking into account order of named axes and broadcast by name! That's semantically cleaner and more future-proof than I expected. (: The thing that I thought would make this impractical is that it's hard optimize misaligned dot product axes (and similar operations) where silent transposes will happen on device. I guess this is not the case, or axis order optimization could (and should) be automated anyway in the future? Very curious how you arrived at this design.

+1 for allowing arrays and operations with named axes outside of `xmap`, i. e. make named axis arrays first-class in jax as suggested above.

**Comment by Bit0r** [comment]
A more powerful implementation is to use first-class dimensions, and [torchdim](https://github.com/facebookresearch/torchdim) uses objects as dimension "variables"

**Comment by Bit0r** [comment]
@apaszke Perhaps it could be further independent of axis position? By utilizing the `named tensor` feature, operations that do not depend on axis position can be achieved. 

1. For example, the for loop operation can directly specify which axis to loop on, and the framework automatically advances the axis to the first dimension. The entire operation is transparent, and users do not need to write any additional code. 
2. For example, we can directly use axis names and indices on that axis to access tensors.

```python
# tensor.named_shape={'batch':32, 'time':100, 'hidden':200}

t[{'time':0, 'hidden':0}] = 1000 # Select tensor with axis time 0 and axis hidden 0, and set tensor to 1000 with broadcast.

for t in tensor['time']:
# Jax automatically performs dimension permutation for operations: tensor: batch, time, hidden -> time, batch, hidden
# t.named_shape = {'batch':32, 'hidden':200}
    ...
```

**Comment by kentslaney** [comment]
@Bit0r I built your suggestion into some code I wrote for operating on a `tuple` or `namedtuple` of tensors

https://github.com/kentslaney/umap-cam/blob/7be1b04c509850cca4b192d821f9c6761ec6b218/src/umap-cam/group.py#L192-L203
